services/port_manager/src/main/java/com/futurewei/alcor/portmanager/service/implement/PortServiceImpl.java

```diff
@@ -0,0 +1,421 @@
+*/
+package com.futurewei.alcor.portmanager.service.implement;
+
+
+import com.futurewei.alcor.portmanager.executor.AsyncExecutor;
+import com.futurewei.alcor.portmanager.exception.*;
+import com.futurewei.alcor.portmanager.repo.PortRepository;
+import com.futurewei.alcor.portmanager.restwrap.IpAddressRestWrap;
+import com.futurewei.alcor.portmanager.restwrap.MacAddressRestWrap;
+import com.futurewei.alcor.portmanager.restwrap.VpcRestWrap;
+import com.futurewei.alcor.portmanager.rollback.*;
+import com.futurewei.alcor.portmanager.service.PortService;
+import com.futurewei.alcor.web.entity.*;
+import com.futurewei.alcor.web.rest.IpAddressRest;
+import com.futurewei.alcor.web.rest.MacAddressRest;
+import com.futurewei.alcor.web.rest.VpcRest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.stereotype.Service;
+
+import java.util.*;
+import java.util.concurrent.CompletableFuture;
+
+@Service
+@ComponentScan(value="com.futurewei.alcor.web.rest")
+public class PortServiceImpl implements PortService {
+    private static final Logger LOG = LoggerFactory.getLogger(PortServiceImpl.class);
+
+    @Autowired
+    private PortRepository portRepository;
+
+    @Autowired
+    private IpAddressRest ipAddressRest;
+
+    @Autowired
+    private MacAddressRest macAddressRest;
+
+    @Autowired
+    private VpcRest vpcRest;
+
+
+    private void getDefaultSecurityGroup(PortState portState) {
+        //FIXME: send get tenant default security group
+        String defaultSgId =  "tenant-default-security-group-id";
+
+        List<String> securityGroups = new ArrayList<>();
+        securityGroups.add(defaultSgId);
+
+        portState.setSecurityGroups(securityGroups);
+    }
+
+    private void verifySecurityGroup(PortState portState) {
+        List<String> securityGroups = portState.getSecurityGroups();
+        String tenantId = portState.getTenantId();
+
+        for (String securityGroup: securityGroups) {
+            //FIXME: send verify tenant security group
+        }
+    }
+
+    private void unbindSecurityGroups(PortState portState) {
+
+    }
+
+    private void bindSecurityGroups(PortState portState) {
+        verifySecurityGroup(portState);
+
+        //FIXME: Not support yet
+    }
+
+    private void rollBackAllOperations(Stack<PortStateRollback> rollbacks)
+            throws Exception {
+        while (!rollbacks.isEmpty()) {
+            rollbacks.pop().doRollback();
+        }
+    }
+
+    private HostState getHostState(String hostId) {
+        return null;
+    }
+
+    private void addPortToHost(String hostId) {
+        HostState hostState = getHostState(hostId);
+
+        //FIXME: Add port to Host
+    }
+
+    private void tryCreatePortState(PortState portState, Stack<PortStateRollback> rollbacks) throws Exception {
+        //Verify VPC ID
```

Comment:
We should verify Subnet ID, instead of Vpc Id here.

Reply:
But port may not have a Subnet ID. In this case, we need to find a suitable subnet for port, and assign an ip address to port from that subnet.

Reply:
@xieus  But port may not have a Subnet ID. In this case, we need to find a suitable subnet for port, and assign an ip address to port from that subnet.

Reply:
Correct. We have multiple cases to cover, most of the cases would require to verify subnet ID. 

https://docs.openstack.org/api-ref/network/v2/?expanded=create-port-detail#ports

(1) If a user specifies a subnet ID in the optional fixed_ips field, then we have to verify that subnet Id
(2) If the user doesn't give any fixed_ips, we handle it as you suggested.

Reply:
@xieus  For the first case, I think we need to get the range ID by  Subnet ID (At this point, the Subnet ID will be verified), and then assign an ip address from the ip range. At present, Subnet manager have not provided an interface to obtain the Range ID by the Subnet ID. I discussed this with kevin, and he said he would think about it.

Reply:
I think we could make this verification simple by calling the regular Get /subnets/{subnet_id} and the response includes the range ids. @kevin-zhonghao

Reply:
That would be nice，But for the second case, how do i get a suitable Range ID?

Reply:
- Simple solution: random pick one subnet 
  (subnet may run out of ips so need retry a few times if such a case is hit)
- Comprehensive solution: Pick subnet based on ip utilization.
We have the ip utilization data for ip range, we could possibly populate that info to VPC manager and pick a subnet with more available ips.

Reply:
@xieus To reduce interaction with other microservices, we can add an interface to ip manager. The parameter of this interface is that Vpc ID. This interface will selects an ip range based on Vpc ID and assigns an ip address from that ip range. Is that okay?

Reply:
@chenpiaoping Does IP manager has the VPC id?

Reply:
No, not yet.

Reply:
Okay the db key is range id in IP Manager, and subnet ID is part of the value. If we need to VPC Id there, we expect to see inefficient query if we try to get range by vpc id. Any other thoughts?

Reply:
@chenpiaoping Regarding the API to query RangeId by SubnetId, the current Subnet Manager supports that. 

You could use the API 

__GET /project/{projectId}/subnets/{subnetId}__

__Sample Response:__
{
  "subnet": {
    "availability_zone": "string",
    "cidr": "string",
    "description": "string",
    "dhcp_enable": true,
    "dns_list": [
      "string"
    ],
    "gateway_ip": "string",
    "id": "string",
    "ipv4_range_id": "string",
    "ipv6_range_id": "string",
    "mac_address": "string",
    "name": "string",
    "primary_dns": "string",
    "project_id": "string",
    "routes": [
      {
        "associatedTableId": "string",
        "associatedType": "MAIN",
        "description": "string",
        "destination": "string",
        "id": "string",
        "name": "string",
        "priority": 0,
        "project_id": "string",
        "target": "string"
      }
    ],
    "secondary_dns": "string",
    "vpc_id": "string"
  }
}

where ipv4_range_id and ipv6_range_id are what you are looking for :-)

Reply:
Thanks, I found this API. I've submitted a new commit for this.

Reply:
In order to avoid the problem of inefficient query, another idea is that we can create a table to store the mapping between vpc id and rang id, but we need to write this table at the same time when we create or delete an ip range.

Reply:
One discipline is that we follow the same workflows for both case (1) and case (2).

You could either go to SubnetManager to verify subnetId or vpcId, retrieve the range id, and then go to IP manager,
OR, build a mapping in IP manager from VpcId to rangeID as well as SubnetId to RangeId, and only need go to IP manager.

Reply:
Yes, for the case of going to SubnetManager to retrieve a range id,  Since SubnetManager do not know whether there is an available ip address in the range, we may need to go to SubnetManager many times to obtain a available range.

Reply:
@xieus Ip manager provides an interface that randomly assigns ip addresses according to vpcId. Is this solution okay? I'm ready to develop.

Reply:
@chenpiaoping Rethinking the issue, we might have to pick the option of " go to SubnetManager to verify subnetId or vpcId, retrieve the range id, and then go to IP manager". 

The reason is that PortManager needs to visit both RouteManager and IPManager to retrieve routes and to allocate ip, respectively. SubnetManger is the source of truth to retrieve the routes under the subnet and to get the ip range id under the same subnet. 

Therefore, the workflow is like this: 
1. Port Manager => SubnetManager
   1.1 If Subnet Id is known, send and verify subnet Id, retrieve routes under the subnet (as SubnetManager has the latest routes) and to get range Id
   1.2 Otherwise, send vpc id to SubnetManager, SubnetManager picks one subnet, and the rest is the same as Step 1.1

2. Port Manager => Ip Manager, send the range id from Step 1, and allocate ip from the given range.

Currently, Step 1.1 needs a new API from SubnetManager. @kevin-zhonghao and I will work to create one in a parallel PR (#181).

Reply:
@xieus  I have submitted the code, but it is a little different from what you described. Can you review it?